gammapy/makers/reduce.py

```diff
@@ -0,0 +1,187 @@
+        """Make single dataset.
+        Parameters
+        ----------
+        observations : `Observations`
+            Observations
+        """
+
+        observations = np.atleast_1d(observations)
+
+        cutouts_kwargs = self._setup_cutouts(observations)
+        dataset = self.dataset.cutout(**cutouts_kwargs, name=self.dataset.name)
+
+        for observation in observations:
+            log.info(f"Computing dataset for observation {observation.obs_id}")
+            dataset_obs = dataset.copy(name=f"obs-{observation.obs_id}_{dataset.name}")
+            for maker in self.makers:
+                log.info(f"Running {maker.tag}")
+                dataset_obs = maker.run(dataset=dataset_obs, observation=observation)
+            if self.write_all:
+                self._write(dataset_obs)
+            dataset.stack(dataset_obs)
+        return dataset
+
+    def _setup_cutouts(self, observations):
+        """set cutout parameters"""
+        if self.cutout_width is None:
+            self.cutout_width = 2 * self.offset_max
+
+        cutout_width = self.cutout_width
+        obs_positions = [obs.pointing_radec.galactic for obs in observations]
+        lon = np.mean([pos.l.value for pos in obs_positions])
+        lat = np.mean([pos.b.value for pos in obs_positions])
+        position = SkyCoord(lon * u.deg, lat * u.deg, frame="galactic")
+        if self.stacking:
+            sep = np.max(position.separation(SkyCoord(obs_positions)))
+            cutout_width += 2 * (sep + 0.1 * u.deg)
+        return {
+            "position": position,
+            "width": cutout_width,
+            "mode": self.cutout_mode,
+        }
+
+    def _write(self, dataset):
+        """Write individual dataset"""
+
+        filename = self.path / f"dataset.fits"
+        log.info(f"Writing {filename}")
+        dataset.write(filename, overwrite=self.overwrite)
+
+    def run(self, observations):
+        """Run and write
+        Parameters
+        ----------
+        observations : `Observations`
+            Observations
+        """
+
+        if self.n_jobs is not None and self.n_jobs > 1:
+            if self.stacking:
+                observations = self._group_observations(self.n_jobs, observations)
+            with contextlib.closing(Pool(processes=self.n_jobs)) as pool:
+                log.info("Using {} jobs.".format(self.n_jobs))
+                results = pool.map(self.make_dataset, observations)
+            pool.join()
+        else:
+            results = [self.make_dataset(obs) for obs in observations]
+
+        if self.stacking:
+            for dataset in results:
+                self.dataset.stack(dataset)
+            self._write(self.dataset)
+            return self.dataset
+
+    @staticmethod
+    def _group_observations(n_groups, observations):
+        """Splint obsevration in multiple groups for parallel run """
+        # TODO: maybe we could introduce observations.select and observations.group
```

Comment:
Thanks @QRemy! Just one preliminary comment here:

Are you sure this "recursive stacking" is required here? From a quick profiling of the `analysis_2.ipynb` notebook it seems we spend 30-40 times more time in computing datasets, then we spend time in stacking them. In this case a simple "sequential" stacking of the datasets, as they "come in" from the parallelised computation should be sufficient and will be much simpler...maybe you can check once if this is true.

However I fully agree with the `TODO` here. Independent of the whether the functionality is used here I think observation selection and grouping is useful in general and ideally it lives on the `Observations` object...

Reply:
The "recursive stacking" is useful if to avoid keeping all the individual datasets in memory, for surveys it is necessary.

Reply:
I pushed a commit using apply_async with callback instead. But I haven't tested the difference in performance yet.